Index: dyncall/dyncall_call_x64_masm.asm
===================================================================
--- dyncall/dyncall_call_x64_masm.asm	(revision 563)
+++ dyncall/dyncall_call_x64_masm.asm	(working copy)
@@ -49,8 +49,7 @@
 
 	sub			rsp, rcx					; Setup stack frame by subtracting the size of the arguments.
 
-	mov			rax, rcx					; Align stack.
-	add			rax, 8
+	mov			rax, rsp					; Align stack.
 	and			rax, 15
 	sub			rsp, rax
 
