Index: /Users/ochafik/src/dyncall/dyncall/dynload/dynload_syms_pe32.c
===================================================================
--- /Users/ochafik/src/dyncall/dyncall/dynload/dynload_syms_pe32.c	(revision 559)
+++ /Users/ochafik/src/dyncall/dyncall/dynload/dynload_syms_pe32.c	(working copy)
@@ -67,7 +67,7 @@
   IMAGE_EXPORT_DIRECTORY* pExports        = (IMAGE_EXPORT_DIRECTORY*) (ptr + pExportsDataDir->VirtualAddress);  
   return pExports->NumberOfNames;
 */
-  return pResolver->count;
+  return (int)pResolver->count;
 }
 
 const char* dlSymsName(DLSyms* pResolver, int index)
@@ -98,3 +98,6 @@
   return (void*) ( pResolver->pBase + pResolver->pFuncs[index] );
 }
 
+/*void* dlSymsNameValue(DLSyms* pSyms, int index, const char** nameOut) {
+	return (*nameOut = dlSymsName(pSyms, index)) ? dlSymsValue(pSyms, index) : NULL;
+}*/
Index: /Users/ochafik/src/dyncall/dyncall/dynload/dynload_darwin.c
===================================================================
--- /Users/ochafik/src/dyncall/dyncall/dynload/dynload_darwin.c	(revision 0)
+++ /Users/ochafik/src/dyncall/dyncall/dynload/dynload_darwin.c	(revision 0)
@@ -0,0 +1,71 @@
+/*
+
+ Copyright (c) 2007-2009 Olivier Chafik <olivier.chafik@gmail.com>
+
+ Permission to use, copy, modify, and distribute this software for any
+ purpose with or without fee is hereby granted, provided that the above
+ copyright notice and this permission notice appear in all copies.
+
+ THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+*/
+
+/*
+
+  dynload_darwin.c
+
+  dynload module for .dylib (mach-o darwin/OS X) files
+
+*/
+
+
+#include "dynload.h"
+
+#include <dlfcn.h>
+
+struct DLLib_
+{
+	char* libPath;
+	void* handle;
+};
+
+DLLib* dlLoadLibrary(const char* libPath)
+{
+	void* handle;
+	size_t len;
+	DLLib* lib;
+	
+	handle = dlopen(libPath, RTLD_LAZY);
+	if (!handle)
+		return NULL;
+	
+	len = strlen(libPath);
+	lib = (DLLib*)malloc(sizeof(DLLib));
+	lib->libPath = (char*)malloc(len + 1);
+	strcpy(lib->libPath, libPath);
+	lib->libPath[len] = '\0';
+	lib->handle = handle;
+	return lib;
+}
+
+void* dlFindSymbol(DLLib* libHandle, const char* symbol)
+{
+  return dlsym(libHandle && libHandle->handle ? libHandle->handle : RTLD_DEFAULT, symbol);
+}
+
+void  dlFreeLibrary(DLLib* libHandle)
+{
+	if (!libHandle)
+		return;
+	
+	dlclose(libHandle->handle);
+	free(libHandle->libPath);
+	free(libHandle);
+}
+
Index: /Users/ochafik/src/dyncall/dyncall/dynload/dynload.h
===================================================================
--- /Users/ochafik/src/dyncall/dyncall/dynload/dynload.h	(revision 559)
+++ /Users/ochafik/src/dyncall/dyncall/dynload/dynload.h	(working copy)
@@ -61,6 +61,7 @@
 int         dlSymsCount  (DLSyms* pSyms);
 const char* dlSymsName   (DLSyms* pSyms, int index);
 void*       dlSymsValue  (DLSyms* pSyms, int index);
+//void*		dlSymsNameValue(DLSyms* pSyms, int index, const char** nameOut);
 
 #ifdef __cplusplus
 }
Index: /Users/ochafik/src/dyncall/dyncall/dynload/dynload_syms.c
===================================================================
--- /Users/ochafik/src/dyncall/dyncall/dynload/dynload_syms.c	(revision 559)
+++ /Users/ochafik/src/dyncall/dyncall/dynload/dynload_syms.c	(working copy)
@@ -23,11 +23,10 @@
 # include "dynload_syms_pe32.c"
 #elif defined(DC_UNIX)
 # if defined (DC__OS_Darwin)
-void dummy() { }
+# include "dynload_syms_darwin.c"
 # else
 # include "dynload_syms_elf.c"
 # endif
 #else
 void dummy() { }
 #endif
-
Index: /Users/ochafik/src/dyncall/dyncall/dynload/dynload_syms_darwin.c
===================================================================
--- /Users/ochafik/src/dyncall/dyncall/dynload/dynload_syms_darwin.c	(revision 0)
+++ /Users/ochafik/src/dyncall/dyncall/dynload/dynload_syms_darwin.c	(revision 0)
@@ -0,0 +1,91 @@
+/*
+
+ Copyright (c) 2007-2009 Olivier Chafik <olivier.chafik@gmail.com>
+
+ Permission to use, copy, modify, and distribute this software for any
+ purpose with or without fee is hereby granted, provided that the above
+ copyright notice and this permission notice appear in all copies.
+
+ THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+*/
+
+/*
+ 
+ dynamic symbol resolver for Mach-O
+
+ */
+
+#include "dynload.h"
+#include "dynload_macros.h"
+#include <mach-o/dyld.h>
+#include <stdio.h>
+
+void dummy() { }
+
+struct DLLib_
+{
+	char* libPath;
+	void* handle;
+};
+
+struct DLSyms_
+{
+	DLLib* pLib;
+	NSObjectFileImage pObjectFileImage;
+};
+
+size_t dlSyms_sizeof()
+{
+  return sizeof(DLSyms);
+}
+void dlSymsInit(DLSyms* pSyms, DLLib* pLib)
+{
+	//const char* path = "/Users/ochafik/nativelibs4java/Runtime/BridJ/src/test/resources/darwin_universal/libtest.dylib";
+	pSyms->pLib = pLib;
+	if (NSCreateObjectFileImageFromFile(//path,
+		pLib->libPath, 
+		&pSyms->pObjectFileImage) != NSObjectFileImageSuccess)
+		pSyms->pObjectFileImage = 0;
+}
+
+void dlSymsCleanup(DLSyms* pSyms)
+{
+	if (pSyms->pObjectFileImage)
+		NSDestroyObjectFileImage(pSyms->pObjectFileImage);
+}
+
+int dlSymsCount(DLSyms* pSyms)
+{
+	if (!pSyms->pObjectFileImage)
+		return 0;
+	return NSSymbolDefinitionCountInObjectFileImage(pSyms->pObjectFileImage);
+}
+
+const char* dlSymsName(DLSyms* pSyms, int index)
+{
+	if (!pSyms->pObjectFileImage)
+		return NULL;
+	return NSSymbolDefinitionNameInObjectFileImage(pSyms->pObjectFileImage, index);
+}
+
+void* dlSymsValue(DLSyms* pSyms, int index)
+{
+	const char* name = dlSymsName(pSyms, index);
+	if (!name)
+		return NULL;
+	return dlFindSymbol(pSyms->pLib, name);
+}
+/*void* dlSymsNameValue(DLSyms* pSyms, int index, const char** nameOut) {
+	if (!pSyms->pObjectFileImage)
+		return false;
+	
+	*nameOut = NSSymbolDefinitionNameInObjectFileImage(pSyms->pObjectFileImage, index);
+	return *nameOut ? dlFindSymbol(pSyms->pLib, *nameOut) : NULL;
+}*/
Index: /Users/ochafik/src/dyncall/dyncall/dynload/dynload.c
===================================================================
--- /Users/ochafik/src/dyncall/dyncall/dynload/dynload.c	(revision 559)
+++ /Users/ochafik/src/dyncall/dyncall/dynload/dynload.c	(working copy)
@@ -21,6 +21,10 @@
 #if defined(DC_WINDOWS)
 #include "dynload_win32.c"
 #elif defined(DC_UNIX)
-#include "dynload_unix.c"
+# if defined (DC__OS_Darwin)
+# include "dynload_darwin.c"
+# else
+# include "dynload_unix.c"
+# endif
 #endif
 
Index: /Users/ochafik/src/dyncall/dyncall/dynload/dynload_syms_elf.c
===================================================================
--- /Users/ochafik/src/dyncall/dyncall/dynload/dynload_syms_elf.c	(revision 559)
+++ /Users/ochafik/src/dyncall/dyncall/dynload/dynload_syms_elf.c	(working copy)
@@ -388,3 +388,6 @@
 }
 #endif
 
+/*void* dlSymsNameValue(DLSyms* pSyms, int index, const char** nameOut) {
+	return (*nameOut = dlSymsName(pSyms, index)) ? dlSymsValue(pSyms, index) : NULL;
+}*/
Index: /Users/ochafik/src/dyncall/dyncall/dynload/dynload_unix.c
===================================================================
--- /Users/ochafik/src/dyncall/dyncall/dynload/dynload_unix.c	(revision 559)
+++ /Users/ochafik/src/dyncall/dyncall/dynload/dynload_unix.c	(working copy)
@@ -37,7 +37,7 @@
 
 void* dlFindSymbol(DLLib* libHandle, const char* symbol)
 {
-  return dlsym( (void*) libHandle, symbol);
+  return dlsym(libHandle ? libHandle : RTLD_DEFAULT, symbol);
 }
 
 void  dlFreeLibrary(DLLib* libHandle)
Index: /Users/ochafik/src/dyncall/dyncall/dyncallback/GNUmakefile
===================================================================
--- /Users/ochafik/src/dyncall/dyncall/dyncallback/GNUmakefile	(revision 559)
+++ /Users/ochafik/src/dyncall/dyncall/dyncallback/GNUmakefile	(working copy)
@@ -60,6 +60,7 @@
 UNITS += dyncall_callback_x86_masm
 endif
 ifdef BUILD_ARCH_x64
+UNITS += dyncall_callback_x64
 UNITS += dyncall_callback_x64_masm
 endif
 endif
Index: /Users/ochafik/src/dyncall/dyncall/dyncallback/dyncall_callback_ppc32.c
===================================================================
--- /Users/ochafik/src/dyncall/dyncall/dyncallback/dyncall_callback_ppc32.c	(revision 559)
+++ /Users/ochafik/src/dyncall/dyncall/dyncallback/dyncall_callback_ppc32.c	(working copy)
@@ -48,6 +48,6 @@
 
 void dcbFreeCallback(DCCallback* pcb)
 {
-  dcbFreeWX(pcb, sizeof(DCCallback));
+  dcFreeWX(pcb, sizeof(DCCallback));
 }
 
